src/dtls_combine.rs

```diff
@@ -0,0 +1,219 @@
+        return Ok((&[], None));
+    }
+
+    let msg_type = first_handshake.msg_type;
+    let message_seq = first_handshake.message_seq;
+    let length = first_handshake.length;
+
+    #[allow(clippy::comparison_chain)]
+    if max > length {
+        // Error case 4
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    } else if max < length {
+        // We do not have all fragments yet
+        return Ok((&[], None));
+    }
+
+    // Write the header into output.
+    {
+        out[0] = msg_type.into(); // The type.
+        out[1..4].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit length
+        out[4..6].copy_from_slice(&message_seq.to_be_bytes()); // 16 bit message sequence
+        out[6..9].copy_from_slice(&[0, 0, 0]); // 24 bit fragment_offset, which is 0 for the entire message.
+        out[9..12].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit fragment_length, which is entire length.
+    }
+
+    let data = &mut out[MESSAGE_HEADER_OFFSET..];
+
+    if data.len() < length as usize {
+        // Error case 1
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    }
+
+    // Loop the fragments, in order and output the data.
+    for handshake in ordered.iter() {
+        if msg_type != handshake.msg_type {
+            // Error case 2
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.length != length {
+            // Error case 3
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.message_seq != message_seq {
+            // Error case 6
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        let from = handshake.fragment_offset as usize;
+        let to = from + handshake.fragment_length as usize;
+
+        let body = match &handshake.body {
+            DTLSMessageHandshakeBody::Fragment(v) => v,
+            _ => {
+                // Error case 5
+                return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+            }
+        };
+
+        // Copy into output.
+        data[from..to].copy_from_slice(&body[..]);
+    }
+
+    // This parse should succeed now and produce a complete message.
+    let (rest, message) = parse_dtls_message_handshake(out)?;
+
+    Ok((rest, Some(message)))
+}
+
+struct Ordered<'a, 'b>([usize; MAX_FRAGMENTS], &'a [DTLSMessageHandshake<'b>]);
+
+impl<'a, 'b> Ordered<'a, 'b> {
+    fn new(fragments: &'a [DTLSMessageHandshake<'b>]) -> Self {
+        // Indexes that will point into handshakes
+        let mut order: [usize; MAX_FRAGMENTS] = array::from_fn(|i| i);
+
+        // Sort the index for fragment_offset starts.
+        order.sort_by_key(|idx| {
+            fragments
+                .get(*idx)
+                .map(|h| h.fragment_offset)
+                // Somewhere outside the fragments length.
+                .unwrap_or(*idx as u32 + 1000)
```

Comment:
What is this value? It seems dangerous to return a random offset if the fragment with this ID does not exist.
Should this return an error? Or is this a magic value?

Also, note that `Option` has a method `map_or_else` that could be used.

Reply:
Yeah. Poor logic here. It's another invariant. `idx` comes from `order` which means it's impossible for `.get(*idx)` to return `None`. Probably better to change this to a direct array reference `fragments[idx].fragment_offset`

I fix.

Reply:
This is a consequence of not using a sorted list.
Well, even if there is an invariant, it is not described (and a minimum would be to use `assert/debug_assert` to state it).